api/query/web_features_manifest_cache.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2024 The WPT Dashboard Project. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package query
+
+import (
+	"context"
+	"sync"
+
+	"github.com/web-platform-tests/wpt.fyi/shared"
+)
+
+// SetWebFeaturesDataCache safely swaps the data cache.
+// Currently, the separate goroutine in the poll folder will use this.
+func SetWebFeaturesDataCache(newData shared.WebFeaturesData) {
+	webFeaturesDataCacheLock.Lock()
+	defer webFeaturesDataCacheLock.Unlock()
+	webFeaturesDataCache = newData
+}
+
+// GetWebFeaturesDataCache safely retrieves the data cache.
+func GetWebFeaturesDataCache() shared.WebFeaturesData {
+	webFeaturesDataCacheLock.RLock()
```

Comment:
I don't see a reason using read lock here, but I guess it is a standard implementation from what I read.

Reply:
The polling services (`StartMetadataPollingService` and `StartWebFeaturesManifestPollingService`) both write to their respective in-memory cache (`query.MetadataMapCached` and `query.webFeaturesDataCache`). And requests to searchcache potentially read from that. There is a potential data race that can occur with the existing metadata cache. By using the lock around the web features cache, it prevents that.

We need to add a similar lock to the MetadataMapCached.

I added some race tests to help exercise the data race to this example PR https://github.com/web-platform-tests/wpt.fyi/pull/3674

Reply:
Aha thanks for the example. I think I didn't exactly distinguish the difference between data race and race condition, where data race pertains to the memory access conflict. Appreciate it!